src/neoxp/Node/Modules/ExpressRpcMethods.cs

```diff
@@ -403,6 +405,90 @@ public JObject GetNep11Properties(JArray @params)
             }
             return json;
         }
+        
+        private const byte Prefix_Contract = 8;
+        private const byte Prefix_NextAvailableId = 15;
+
+        private int LastUsedContractId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.TryGet(key);
+            return (int)(BigInteger)item;
+        }
+
+        private void SetLastUsedContractId(DataCache snapshot, int newId)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            item.Set(newId);
+        }
+        
+        private int GetNextAvailableId(DataCache snapshot)
+        {
+            StorageKey key = new KeyBuilder(NativeContract.ContractManagement.Id, Prefix_NextAvailableId);
+            StorageItem item = snapshot.GetAndChange(key);
+            int value = (int)(BigInteger)item;
+            item.Add(1);
+            return value;
+        }
+        
+        [RpcMethod]
+        public JObject? ExpressPersistContract(JObject @params)
+        {
+            var state = @params[0]["state"];
+            var storagePairs = (JArray) @params[0]["storage"];
+            var snapshot = neoSystem.GetSnapshot();
+
+            var nef = new NefFile
+            {
+                CheckSum = (uint)state["nef"]["checksum"].AsNumber(),
+                Compiler = state["nef"]["compiler"].AsString(),
+                Source = state["nef"]["source"].AsString(),
+                Tokens = new MethodToken[0], // TODO: support parsing MethodTokens or use fix of https://github.com/neo-project/neo/issues/2674
+                Script = Convert.FromBase64String(state["nef"]["script"].AsString())
+            };
+            
+            ContractManifest m = ContractManifest.FromJson(state["manifest"]);
+            
+            // Our local chain might already be using the contract id of the pulled contract, we need to check for this
+            // to avoid having contracts with duplicate id's. This is important because the contract id is part of the
+            // StorageContext used with Storage syscalls and else we'll potentially override storage keys or iterate
+            // over keys that shouldn't exist for one of the contracts.
+            int contractId = (int)state["id"].AsNumber(); 
+            if (contractId <= LastUsedContractId(snapshot))
+            {
+                contractId = GetNextAvailableId(snapshot);
+            }
+            else
+            {
+                // Update available id such that a regular contract deploy will use the right next id;
+                SetLastUsedContractId(snapshot, contractId);
+            }
+            
+            var c = new ContractState
```

Comment:
`RpcClient` has a `ContractStateFromJson` method. Why aren't you using that?

Reply:
Didn't know that existed. I even created an issue https://github.com/neo-project/neo/issues/2674 to get it natively supported :sweat_smile:

Reply:
I would argue that all the from/to JSON methods should be in the core neo.dll not in modules, but since express already references RpcClient assembly anyway, no big deal.

Reply:
:100: there's quite a few places where the API in the core is asymmetrical exposing just with what they need, not what others might need.